Mathlib/Computability/TuringDegrees.lean

```diff
@@ -0,0 +1,355 @@
+- `turing_equivalent`: A relation defining Turing equivalence between partial functions.
+- `TuringDegree`:
+  The type of Turing degrees, defined as equivalence classes under `turing_equivalent`.
+- `join`: Combines two partial functions into one by =
+  mapping even and odd numbers to respective functions.
+
+## Notation
+
+- `f ≤ᵀ g` : `f` is Turing reducible to `g`.
+- `f ≡ᵀ g` : `f` is Turing equivalent to `g`.
+- `f ⊕ g` : The join of partial functions `f` and `g`.
+
+## Implementation Notes
+
+The type of partial functions recursive in an oracle `g` is the smallest type containing
+the constant zero, the successor, projections, the oracle `g`, and is closed under
+pairing, composition, primitive recursion, and μ-recursion. The `join` operation
+combines two partial functions into a single partial function by mapping even and odd
+numbers to the respective functions.
+
+## References
+
+* [Carneiro2018] Carneiro, Mario.
+  *Formalizing Computability Theory via Partial Recursive Functions*.
+  arXiv preprint arXiv:1810.08380, 2018.
+* [Odifreddi1989] Odifreddi, Piergiorgio.
+  *Classical Recursion Theory: The Theory of Functions and Sets of Natural Numbers,
+  Vol. I*. Springer-Verlag, 1989.
+* [Soare1987] Soare, Robert I. *Recursively Enumerable Sets and Degrees*. Springer-Verlag, 1987.
+* [Gu2015] Gu, Yi-Zhi. *Turing Degrees*. Institute for Advanced Study, 2015.
+
+## Tags
+
+Computability, Oracle, Turing Degrees, Reducibility, Equivalence Relation
+-/
+
+open Primrec Nat.Partrec Part
+
+/--
+The type of partial functions `f` that are recursive in an oracle `g` is the smallest type
+containing the constant zero, the successor, left and right projections, the oracle `g`,
+and is closed under pairing, composition, primitive recursion, and μ-recursion.
+
+Equivalently one can say that `f` is turing reducible to `g` when `f` is recursive in `g`.
+-/
+inductive RecursiveIn : (ℕ →. ℕ) → (ℕ →. ℕ) → Prop
+  | zero {g} : RecursiveIn (fun _ => 0) g
+  | succ {g} : RecursiveIn Nat.succ g
+  | left {g} : RecursiveIn (fun n => (Nat.unpair n).1) g
+  | right {g} : RecursiveIn (fun n => (Nat.unpair n).2) g
+  | oracle {g} : RecursiveIn g g
+  | pair {f h g : ℕ →. ℕ} (hf : RecursiveIn f g) (hh : RecursiveIn h g) :
+      RecursiveIn (fun n => (Nat.pair <$> f n <*> h n)) g
+  | comp {f h g : ℕ →. ℕ} (hf : RecursiveIn f g) (hh : RecursiveIn h g) :
+      RecursiveIn (fun n => h n >>= f) g
+  | prec {f h g : ℕ →. ℕ} (hf : RecursiveIn f g) (hh : RecursiveIn h g) :
+      RecursiveIn (fun p =>
+        let (a, n) := Nat.unpair p
+        n.rec (f a) (fun y ih => do
+          let i ← ih
+          h (Nat.pair a (Nat.pair y i)))) g
+  | rfind {f g : ℕ →. ℕ} (hf : RecursiveIn f g) :
+      RecursiveIn (fun a =>
+        Nat.rfind (fun n => (fun m => m = 0) <$> f (Nat.pair a n))) g
+
+
+/--
+`f` is Turing equivalent to `g` if `f` is reducible to `g` and `g` is reducible to `f`.
+-/
+def turing_equivalent (f g : ℕ →. ℕ) : Prop :=
```

Comment:
```suggestion
def TuringEquivalent (f g : ℕ →. ℕ) : Prop :=
```
as this is a `Prop`. The lemmas about it should either be called `TuringEquivalent.foo` or `turingEquivalent_foo`.